### DISCUSSION TITLE
args_parse function may have a bug

### DISCUSSION THREAD

**Discussion by zxxunkxzhx** [VirusTotal/yara#1652] [General]
case ARGS_OPT_INTEGER:
    if (opt_arg == NULL)
      return ARGS_ERROR_REQUIRED_INTEGER_ARG;

    *(long*) opt->value = _tcstol(opt_arg, &endptr, 0);

    if (*endptr != '\0')  ////////////////////////////////// endptr point to the last non-digit char, it may NOT point to '\0'
      return ARGS_ERROR_REQUIRED_INTEGER_ARG;

    if (opt_arg_was_used != NULL)
      *opt_arg_was_used = 1;

    break;

**Comment by plusvic** [comment]
Exactly, `endptr` always point to the ending null-character (`\0`) if the input is all digits. If `endptr` points to something else is because the input contains some non-digit character. In such cases it returns `ARGS_ERROR_REQUIRED_INTEGER_ARG`. If the user enters something like `123x` it doesn't take the `123` as a number and discard the `x`, it returns an error.
